src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -1871,14 +1873,20 @@ void Session::send_message()
             return false;
         }
 
-        return m_upload_progress.client_version >= m_pending_flx_sub_set->snapshot_version;
+        // Send QUERY messages when the upload progress client version reaches the snapshot version
+        // of a pending subscription, or if this is a fresh realm download session, since UPLOAD
+        // messages are not allowed and the upload progress will not be updated.
+        return m_upload_progress.client_version >= m_pending_flx_sub_set->snapshot_version ||
+               REALM_UNLIKELY(m_fresh_realm_download);
```

Comment:
In a fresh realm the only uploads we would normally be sending are download acknowledgements. If there was just another branch [here](https://github.com/realm/realm-core/blob/c82d03beac67d4a9447ced1a2e42cc2cb08cd31a/src/realm/sync/noinst/client_impl_base.cpp#L2050) to return early if sending upload acks is disabled, would that be sufficient? Also, is checking at this layer whether this sync::Session is a fresh realm right? Should we be checking that up in the object-store layer as an option when starting the session?

Reply:
it would be enough indeed. we still need to make sure that queries are sent.
>Also, is checking at this layer whether this sync::Session is a fresh realm right?

I think we usually do that. The annoying part is that we need to access that flag through several layers.

Reply:
By the time you reach [here](https://github.com/realm/realm-core/blob/c82d03beac67d4a9447ced1a2e42cc2cb08cd31a/src/realm/sync/noinst/client_impl_base.cpp#L2050) you've already advanced your local upload cursor so sending queries should Just Work Normally™️.

I was under the impression that checking whether a realm is a "fresh" realm happens mostly in object-store? Where else down in sync::Session do we check if the realm ends in ".fresh"?

Reply:
I quoted the wrong sentence.

>By the time you reach here you've already advanced your local upload cursor so sending queries should Just Work Normally™️.

I think the idea is to not advance the upload cursor at all. I'm checking to see if that's actually really needed.

Reply:
Advancing the upload cursor is what got us into `Last integrated client version on server cannot be greater than the latest client version in existence` unfortunately.

Reply:
But if we never send an upload message, won't the server send 0, so the server version will never be greater than the local version?

Reply:
by "advancing the local upload cursor", all i really mean is calling [find_uploadable_changesets() here](https://github.com/realm/realm-core/blob/c82d03beac67d4a9447ced1a2e42cc2cb08cd31a/src/realm/sync/noinst/client_impl_base.cpp#L2038). That just updates the in-memory state of what the sync client thinks its already processed, but doesn't necessarily have to send an upload. This is what we already do for empty uploads.

I don't know what you mean by the server sending zero so the server version will never be greater than the local version.

Reply:
I see what you mean now. We advance the upload cursor as if we ack the downloads. I think your suggestion to return early in `send_upload_message()` would work then.